core/iwasm/compilation/aot.c

```diff
@@ -60,27 +63,39 @@ aot_create_mem_init_data_list(const WASMModule *module)
 
     /* Create each memory data segment */
     for (i = 0; i < module->data_seg_count; i++) {
-        size = offsetof(AOTMemInitData, bytes)
-               + (uint64)module->data_segments[i]->data_length;
+        size = sizeof(AOTMemInitData);
         if (size >= UINT32_MAX
             || !(data_list[i] = wasm_runtime_malloc((uint32)size))) {
             aot_set_last_error("allocate memory failed.");
             goto fail;
         }
 
 #if WASM_ENABLE_BULK_MEMORY != 0
+        /* Set bulk memory specific properties if enabled */
         data_list[i]->is_passive = module->data_segments[i]->is_passive;
         data_list[i]->memory_index = module->data_segments[i]->memory_index;
 #endif
         data_list[i]->offset = module->data_segments[i]->base_offset;
         data_list[i]->byte_count = module->data_segments[i]->data_length;
-        memcpy(data_list[i]->bytes, module->data_segments[i]->data,
-               module->data_segments[i]->data_length);
+        data_list[i]->bytes = NULL;
+        /* Allocate memory for AOT compiler is OK, because the data segment
+         * is small and the host memory is enough */
```

Comment:
If that's the case, why not maintain the original? But still, the modification was made.

Reply:
Are you suggesting that malloc()/free() should better match the 
implied convention of the data structure? If using a variable-length array as we did previously, we should allocate one block of memory. If using a pointer as we do now, we should allocate memory twice.

Reply:
Using variable-length arrays is not an issue; it's just that in the current scenario, most of the time we don't even need to allocate memory a second time. We can simply point the pointer to the data buffer.

Reply:
I completely agree the intent of this PR. 

My uncertainty lies with the changes made in aot.c. As you mentioned, the AOT compiler doesn't need to *set the pointer to the data buffer*. From my perspective, this suggests that there is no need to alter aot.c, as it does not align with the same intention as aot_loader.c. But if this is about upholding an implied convention of programming, then I believe it is acceptable.

Reply:
The key point is the structure that `uint8* bytes; // NEW` at the end can not be used with variable-length, it's totally different with `uint8 bytes[1]`, so I must modifty the `aot.c`.

Reply:
Oh, I see.